### PR TITLE
[llvm] Support atomic operations of f16

### PR DESCRIPTION
**Pull request by sjwsl**
Support atomic add/max/min of `f16` by CAS (`CreateAtomicCmpXchg`) on CPU and CUDA backends. 
CPU: Cast to `i16` pointer.
CUDA: Extend to `i32` pointer in a hacky way because CAS for 16-bit values is not supported. The correctness depends on the memory arrangement.

### CODE REVIEW

**Comment by bobcao3** [comment]
oops, vulkan f16 atomics failed.

File: tests/python/test_f16.py [hunk 1]
```diff
@@ -218,3 +218,60 @@ def paint(t: float):
...
+        for i in range(1000):
+            f[0] += 1.12
+
+        # Serial sum
+        for _ in range(1):
+            for i in range(1000):
+                f[1] = f[1] + 1.12
+
+    foo()
+    assert (f[0] == f[1])
```
Comment (bobcao3) [inline tests/python/test_f16.py:239]: This can fail, we need an epsilon when comparing floats as its not commutative

Comment (sjwsl) [reply tests/python/test_f16.py:240]: Thanks!

**Comment by sjwsl** [comment]
> oops, vulkan f16 atomics failed.

Ah guess we have not supported atomic ops on Vulkan? Now I temporarily exclude Vulkan in related tests to pass CI.

**Review by ailzhang — APPROVED** [review]
Awesome!! Thanks!

File: taichi/backends/cuda/codegen_cuda.cpp [hunk 2]
```diff
@@ -330,15 +331,174 @@ class CodeGenLLVMCUDA : public CodeGenLLVM {
         llvm::AtomicOrdering::SequentiallyConsistent);
   }
 
+  // A huge hack for supporting f16 atomic add/max/min! Borrowed from
+  // https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/ir_emitter.cc#L378-L490
```
Comment (k-ye) [inline taichi/backends/cuda/codegen_cuda.cpp:335]: tiny nit: https://github.com/tensorflow/tensorflow/blob/470d58a83470f8ede3beaa584e6992bc71b7baa6/tensorflow/compiler/xla/service/gpu/ir_emitter.cc#L378-L490

Comment (sjwsl) [reply taichi/backends/cuda/codegen_cuda.cpp:335]: Thanks!

**Pull request merged**